How to Learn Dynamics Incoherently

DSGuala opened the pull request:
**Title:**

Learning quantum dynamics incoherently: Variational learning using classical shadows

**Summary:**

This demo describes how to use a parameterized quantum circuit and classical shadow measurements to reproduce an unknown quantum process.

**Relevant references:**

https://arxiv.org/abs/2303.12834

----
If you are writing a demonstration, please answer these questions to facilitate the marketing process.

* GOALS — Why are we working on this now?

  - Promote PennyLane datasets
  - Promote the Learning Dynamics Incoherently dataset
  - Highlight and encourage dataset contributions from external contributors
  - Show PennyLane implementation of a paper


* AUDIENCE — Who is this for?

  QML researchers


* KEYWORDS — What words should be included in the marketing post?


* Which of the following types of documentation is most similar to your file? 
(more details [here](https://www.notion.so/xanaduai/Different-kinds-of-documentation-69200645fe59442991c71f9e7d8a77f8))
    
- [ ] Tutorial
- [ ] Demo
- [x] How-to

Qottmann commented:
Hey @DSGuala please make sure the demo renders when tagging for review. I tried fixing it locally by providing the missing argument but then encounter other errors. The older deploy does not seem to show this new demo but only "Create an Ising Hamiltonian with random weights"

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+# We first create the Hamiltonian. It will be trotterized later when we use it in a quantum circuit.
+
+import pennylane as qml
+from pennylane import numpy as np
```
I would suggest `import pennylane.numpy as pnp` to avoid confusion with regular numpy, there has been some collective effort to mitigate that confusion. I would further suggest to separately import regular numpy and use that for everything, except the trainable parameters. Should make the demo also faster and avoids having to manually set `requires_grad = False` below 👍

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+# reproduced from Understanding the Haar measure:
+def qr_haar(N):
+    """Generate a Haar-random matrix using the QR decomposition."""
+    # Step 1
```
I think its fine to link to the haar random demo, but I think the interested reader would still benefit from brief in-code descriptions of these steps.
Like "Step 1: generate a random complex square matrix"..

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+print(qml.draw(target_circuit)())
+
+n_shadows = 1000
+# shadows=[]
+# for random_state in random_states:
+#     bits, recipes = target_circuit(shots=n_shadows)
+#     shadow = qml.ClassicalShadow(bits, recipes)
+#     shadows.append(shadow)
```
should this be removed?

DSGuala replied:
:thinking: 
I will either use it, remove it, or explain it.

Basically we want to perform this training over several initial states, but in the demo we're only training on a single state. The commented code is to loop over multiple training states.

Qottmann replied:
Ah right! Yeah indeed one input state seems not sufficient

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+# Create model circuit that will learn the target process
+# ---------------------------------------------------------
+#
+# Now that we have the classical shadow measurements, we need to create a `model_circuit` that
```
double ticks in ReST
```suggestion
# Now that we have the classical shadow measurements, we need to create a ``model_circuit`` that
```

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+# learns to produce the same output as the target circuit. We will then use the classical shadow
+# measurements to estimate the similarity between the `model_circuit` and the `target_circuit`. 
+#
+# Knowing that the `target_circuit` uses the :class:`~pennylane.TrotterProduct`, we choose a
+# `model_circuit` that matches the :class:`~pennylane.TrotterProduct` structure. If the target
+# target quantum process were truly unknown, then we could choose a general variational quantum
+# circuit. 
```
Is this what they are doing in the paper as well? It sounds a bit odd to me because when you are using a Trotter product of the same Hamiltonian you already use a lot of information of the "unknown" process 🤔 i.e. your inductive bias is already towards the target model to start with

If that is something that they do in the paper that we just reproduce then thats fine, I would just maybe rephrase some parts to make that clear

DSGuala replied:
Yup, it feels a bit like cheating. But according to the paper they are
>  training a simulated parameterized quantum
circuit (PQC) with the same gate structure as the target Trotterised unitary.

Qottmann replied:
Okay that's quite odd then. I think we should just make sure to make that clear in the demo (i.e. that this the paper's and not our choice, and that it is in some way cheating)

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+    return qml.state()
+
+
+placeholder_trotter_circuit()
+
+ops = qml.compile(placeholder_trotter_circuit.tape)[0][0].operations
```
What's going on here?

DSGuala replied:
This is just a trick to get the trotter circuit to decompose like I want it to. Removed and replaced with adding RX and IsingZZ gates in a loop.

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+
+placeholder_trotter_circuit()
+
+ops = qml.compile(placeholder_trotter_circuit.tape)[0][0].operations
```
Reads a bit awkward to have to compile in order to obtain the ops. Is there not an easier way to do that in PL?

DSGuala replied:
We could run the circuit and then do `circuit.tape.operations` but then we have to decompose them.

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+
+@qml.qnode(dev)
+def model_circuit(params, random_state):
+    # this is a parameterized quantum circuit with the same gate structure as the target trotterised unitary
+    qml.StatePrep(random_state, wires=range(n_qubits))
+    for op, param in zip(ops, params):
+        if op.name == "RX":
+            qml.RX(param, wires=op.wires)
+        elif op.name == "IsingZZ":
+            qml.IsingZZ(param, wires=op.wires)
+    return [qml.density_matrix(i) for i in range(n_qubits)]
```
I am a bit confused here. So you are constructing the Trotter circuit, take in its `ops`, but then only use the RX and IsingZZ bits? Are there other ops in there that you ignore with this? Why not just do blocks of RX and IsingZZ in a unfiorm way?

DSGuala replied:
No operations are removed here. I'm just processing the operations in the list.
I updated this to add RX and IsingZZ in a loop. It's less obvious that it's copying the target circuit structure, but clearer in implementation.

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+def cost(params):
+    cost = 0.0
+    # for random_state in random_states:
+    observable_mats = model_circuit(params, random_states[0])
+    observable_pauli = [qml.pauli_decompose(observable_mat, wire_order=[qubit]) for qubit, observable_mat in enumerate(observable_mats)]
+    cost = cost - qml.math.sum(shadow.expval(observable_pauli))
```
Can you directly measure and use classical shadows in `model_circuit`? Not sure I get what is happening here fully, so might be missing something

DSGuala replied:
The procedure is kind of unusual. We don't use classical shadows in the model circuit. 

We simulate the model circuit to get local density matrices for each qubit.
We then use those density matrices as *observables*. The classical shadows are used to estimate the expectation value of those observables. If there is a lot of overlap between the model output and target output, then the expectation value should be high.

Qottmann replied:
I feel like since in the end we just process pauli measurements in one way or the other this should be formulatable directly as a shadow measurement. Can you expand on the cost function you are computing analytically in the text (would be good for understanding anyway), based on that I might be able to help you with this

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+for i in range(steps):
+    params, final_cost = optimizer.step_and_cost(cost, params)
+
+print("Final cost:", final_cost)
```
Doesnt look like it is optimizing much?
```
Initial cost: -2.176236321833425
Final cost: -2.3809491417733857
```
could you plot the training progress per epochs?

DSGuala replied:
Sure. Here is a plot for the new cost function. The lowest possible cost is 0.39 and we start at 0.51 with random parameters. By optimizing for 1000 iterations (2 minutes) we get to 0.42 and it starts to level off.

Qottmann replied:
Ah nice! Would be good to include that plot, and perhaps plot it as (cost - cost_min) to see how close we are to the minimum 👍

Qottmann replied:
so the output now is
```
Initial cost: 0.5198848497847514
Final cost: 0.45980806723330603
```
That alone is not very insightful, would be great to see the plot here 👍 

It would also be good to elaborate that there is a non-zero minimum of that cost function, where it is, and why it is there

ikurecic replied:
I agree. I was staring at the output here and trying to justify it... this could do the trick, if you could add it.

DSGuala replied:
Added the plot and explained why the ideal parameters give a non-zero cost. It's related to the fact these are mixed states with $Tr(\rho^2)< 1$

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+
+print("Final cost:", final_cost)
+
+print(model_circuit(params, random_states[0]))
```
Was this for you to debug or is this something you want the reader to see? If the latter, what can I read from this?

Same for the local snapshots below 🤔

DSGuala replied:
I will add text to describe what we see here and print something more clearly. I want to show that after optimizing, the mixed state for qubit 0 is the same for the target and model circuits.

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+# ----------------------------------------------------------------
+#
+
+[ds] = qml.data.load("other", name="learning-dynamics-incoherently")
```
thats cool, didnt know that trick!

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+######################################################################
+# Now we repeat using the learning dynamics incoherently dataset
+# ----------------------------------------------------------------
+#
```
some text to describe what is going on here would be great

Qottmann reviewed (commented):
Good work @DSGuala 💪 

The how-to demo gives a concise recipe for reproducing the main results form the quoted paper, that's great 👍 

I left some question regarding some things that strike me as odd and where I am not sure if those were your choices or just reproducing the paper. I think in either case we should rephrase some things to make that more clear to the reader.

The demo currently misses a convincing code part that shows that the learned circuit approximates well the target circuit are. I am not sure that is the case actually given the little improvement in the cost value 🤔 

There were a lot of avoidable ReST vs markdown rendering issues. Please make sure to do a self-review of the rendered version to avoid that in the future 👍

DSGuala commented:
> Good work @DSGuala 💪
> 
> The how-to demo gives a concise recipe for reproducing the main results from the quoted paper, that's great 👍
> 
> I left some question regarding some things that strike me as odd and where I am not sure if those were your choices or just reproducing the paper. I think in either case we should rephrase some things to make that more clear to the reader.
> 
> The demo currently misses a convincing code part that shows that the learned circuit approximates well the target circuit. I am not sure that is the case actually given the little improvement in the cost value 🤔
> 
> There were a lot of avoidable ReST vs markdown rendering issues. Please make sure to do a self-review of the rendered version to avoid that in the future 👍

Thanks for the review! Super useful comments. 

Sorry for the poor review experience :disappointed: I'm aware that this demo still needs a lot of work. Unfortunately, I haven't been able to dedicate as much focus-time as I wanted to it. But I will be addressing all the review comments as soon as possible and do some troubleshooting to make sure the output clearly demonstrates that the circuit is learned.

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+r"""How to learn quantum dynamics incoherently
+==========================================
+
+If we have an unknown quantum process ❓ that takes a quantum state as input and outputs
```
It was intentional, but I'll find a better way to include emojis :+1:

ikurecic replied:
You could try this:
```suggestion
If we have an unknown quantum process (❓) that takes a quantum state as input and outputs
```
But also, why use emoji to denote this if you don't refer to it anywhere going forward. I would just lose it completely.

ikurecic replied:
```suggestion
If we have an unknown quantum process that takes a quantum state as input and outputs
```

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+alphas = np.random.normal(0, 0.5, size=n_qubits)
+hamiltonian = qml.sum(
+    *[qml.PauliZ(wires=i) @ qml.PauliZ(wires=i + 1) for i in range(n_qubits - 1)]
+)+ qml.sum(*[alphas[i] * qml.PauliX(wires=i) for i in range(n_qubits)])
```
Nice catch!

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+# process.
+#
+# Ideally, our input states should be uniformly distributed over the state space. If they are all
+# clustered together, our model circuit will not learn to approximate the "unknown" quantum process
```
Because in our example we technically know what it is, a trotterized hamiltonian. But I'll remove this.

Qottmann commented:
I noticed that the paper's title and focus also is on the limitations of these procedures. Is that something we want to mention / discuss in this how-to as well?

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+4. Create a model variational circuit to learn the quantum process.
+5. Train the variational circuit.
+6. Repeat the procedure using the target quantum process and hardware measurements used in
+`The power and limitations of learning quantum dynamics incoherently <https://arxiv.org/abs/2303.12834>`__
```
You should make this into a citation. You can take the syntax from other demos — and remember to include the references in the metadata, as well.
(If you would rather I take care of this part, let me know. I'd have to contribute to the pull request to not make that process very annoying, so I want to check with you if that's okay.)

DSGuala replied:
Done!
Yup, feel free to push directly to this branch.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+"""
+
+######################################################################
+# Creating the unknown target quantum process
+# --------------------------------------------
```
```suggestion
# Creating an unknown target quantum process
# -------------------------------------------
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+# Creating the unknown target quantum process
+# --------------------------------------------
+# 
+# We can perform a well known quantum process, imaginary time evolution of a Hamiltonian:
```
```suggestion
# To create an unknown quantum process, we can start by performing a well-known quantum process, the imaginary time evolution of a Hamiltonian:
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+# 
+# We can perform a well known quantum process, imaginary time evolution of a Hamiltonian:
+# 
+# .. math:: U(H, t) = e^{-i H t / \hbar}
```
```suggestion
# .. math:: U(H, t) = e^{-i H t / \hbar} .
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+# 
+# .. math:: U(H, t) = e^{-i H t / \hbar}
+#
+# Specifically, we will use the same Hamiltonian as in the paper,
```
```suggestion
# Specifically, here we will use the same Hamiltonian as in the referenced paper,
```

ikurecic replied:
Also include the reference here, please.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+# 
+# We use a Trotterized version of this Hamiltonian through :class:`~pennylane.TrotterProduct`.
+# 
+# We first create the Hamiltonian. It will be trotterized later when we use it in a quantum circuit.
```
```suggestion
# We first create the Hamiltonian. It will be Trotterized later when we use it in a quantum circuit.
```
Consistency

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+#
+# For quantum systems, this means we want to sample `Haar random states <https://en.wikipedia.org/wiki/Haar_measure>`__.
+# We will create Haar random states using a procedure from our demo,
+# [Understanding the Haar measure](https://pennylane.ai/qml/demos/tutorial_haar_measure/):
```
Markdown links don't work here. You could use this:
```suggestion
# :doc:`Understanding the Haar measure <https://pennylane.ai/qml/demos/tutorial_haar_measure/>`:
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+#
+# .. note ::
+#
+#    Using this method to produce random Haar states becomes computationally expensive 
+#    around 10 qubits.
```
You should add some context and/or a reference here.

DSGuala replied:
Rephrased to "On a personal computer, this method becomes slow (>1 second) around 10 qubits." to imply that we tested this and that that's what users should expect on their computers.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+)+ qml.dot(alphas, [qml.PauliX(wires=i) for i in range(n_qubits)])
+
+######################################################################
+# Create random initial states
+# ------------------------------
```
```suggestion
# Creating random initial states
# --------------------------------
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+#
+# The next step in our procedure is to prepare several initial states. We will then apply the
+# "unknown" quantum process to each of these states to create input-output pairs. That is, for each
+# input state, we will be able to measure the output state after appllying the "unknown" quantum
```
```suggestion
# input state, we will be able to measure the output state after applying the "unknown" quantum
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+shadow = qml.ClassicalShadow(bits, recipes)
+
+######################################################################
+# Create model circuit that will learn the target process
+# ---------------------------------------------------------
```
```suggestion
# Creating a model circuit that will learn the target process
# -------------------------------------------------------------
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+
+print(np.mean(shadow.local_snapshots(),axis=0))
+
+######################################################################
+# Now we repeat using the learning dynamics incoherently dataset
+# ----------------------------------------------------------------
```
Really?

ikurecic reviewed (commented):
Hey @DSGuala , you're right, this still needs quite a bit of work. I've left some very minor comments here as I was going through it because it will need a more detailed review later on, but please reach out when you've identified something where I can help you get this into a better shape in the next couple of days. Remember it still needs to go through two stages of review after the technical review is done — these are shorter, but they still require some time.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,292 @@
+# We first create the Hamiltonian. It will be Trotterized later when we use it in a quantum circuit.
+
+import pennylane as qml
+from pennylane import numpy as pnp
```
![image](https://github.com/user-attachments/assets/d2dfe816-1a54-49e7-86c5-54b5b0245f3b)
FYI. Worth changing it all in the code at the same time so the checks pass.

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+r"""How to learn quantum dynamics incoherently
+==========================================
+
+If we have an unknown quantum process that takes a quantum state as input and outputs
+another state, how can we recreate it or simulate it in a quantum circuit? How do we create a
+model circuit that reproduces the target quantum process? One approach is to learn the
+dynamics of this process incoherently [#Jerbi]_. In simple terms, this
+consists of two steps:
```
(optional) I wonder if this simplifies the intro a bit:
```suggestion
If we have an unknown quantum process that takes a quantum state as input and outputs
another state, how can we recreate it or simulate it in a quantum circuit? One approach is to learn the
dynamics of this process incoherently [#Jerbi]_. In simple terms, this
consists of two steps:
```

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+# where :math:`n` is the number of qubits and :math:`\alpha` are randomly generated weights.
+# 
+# We first create the Hamiltonian and Trotterize later in a quantum circuit via
+# :class:`~pennylane.TrotterProduct`_.
```
```suggestion
# :class:`~pennylane.TrotterProduct`.
```

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+#    On a personal computer, this method becomes slow (>1 second) around 10 qubits. 
+#    
+
+from numpy.linalg import qr
+
+# reproduced from Understanding the Haar measure:
+def qr_haar(N):
+    """Generate a Haar-random matrix using the QR decomposition."""
+    # Step 1: generate an N x N matrix Z with normally distributed complex numbers
+    A, B = np.random.normal(size=(N, N)), np.random.normal(size=(N, N))
+    Z = A + 1j * B
+
+    # Step 2: compute a QR decomposition of Z
+    Q, R = qr(Z)
+
+    # Step 3: create a diagonal matrix  
+    Lambda = np.diag([R[i, i] / np.abs(R[i, i]) for i in range(N)])
+
+    # Step 4: a Haar-random matrix Q' = Q@Lambda
+    return np.dot(Q, Lambda)
+
+random_unitary = qr_haar(2**n_qubits)
+
+random_states = [random_unitary[:,0]]
```
@DSGuala could you replace this with just `scipy.stats.unitary_group`? E.g.,

```python
from scipy.stats import unitary_group
random_unitary = unitary_group.rvs(2 ** n_qubits)
random_states = [random_unitary[:, 0]]
```

DSGuala replied:
Not sure :thinking: Is it Haar-random?

josh146 replied:
Oh right 🤦 I forgot about the Haar random aspect.

DSGuala replied:
Looks like they're not Haar-random. Plotting on a bloch sphere shows the samples are concentrated around the middle:
![image](https://github.com/user-attachments/assets/3a79e6c1-ddab-43b1-bc8e-fb210132e805)

josh146 replied:
🫨

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+#
+# Now that we have the classical shadow measurements, we need to create a ``model_circuit`` that
+# learns to produce the same output as the target circuit. We will then use the classical shadow
+# measurements to estimate the similarity between the `model_circuit` and the `target_circuit`. 
```
```suggestion
# measurements to estimate the similarity between the ``model_circuit`` and the ``target_circuit``. 
```

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+#
+# As done in [#Jerbi]_, we create a ``model_circuit`` with the same gate structure as the target
+# structure. If the target quantum process were truly unknown, then we could choose a general
+# variational quantum circuit like in the `Variational classifier demo <https://pennylane.ai/qml/demos/tutorial_variational_classifier/>`_. 
```
```suggestion
# variational quantum circuit like in the :doc:`Variational classifier demo <demos/tutorial_variational_classifier>`. 
```

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+# 5. Training a model circuit using the classical shadows in a cost function
+# ---------------------------------------------------------------------------
+#
+# We now have to find the optimal parameters for `model_circuit` to mirror the `target_circuit`. 
```
```suggestion
# We now have to find the optimal parameters for ``model_circuit`` to mirror the ``target_circuit``. 
```

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+# Using the learning dynamics incoherently dataset
+# ----------------------------------------------------------------
+#
+# [#Jerbi]_ Performs the procedure described above on a larger, 16-qubit transverse field Ising
```
I would replace the autoreference here with text, e.g., 
```suggestion
# In Jerbi et al. [#Jerbi]_, they performs the procedure described above on a larger, 16-qubit transverse field Ising
```

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+#
+# [#Jerbi]_ Performs the procedure described above on a larger, 16-qubit transverse field Ising
+# Hamiltonian and uses classical shadow samples from quantum hardware to estimate the cost function.
+# This data is available in PennyLane through the :class:`qml.data`_ module. More information about
```
```suggestion
# This data is available in PennyLane through the :class:`qml.data` module. More information about
```

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+r"""How to learn quantum dynamics incoherently
+==========================================
+
+If we have an unknown quantum process that takes a quantum state as input and outputs
```
I recommend mentioning/linking the dataset page in the first paragraph, as it is something we want to advertise!

josh146 commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,306 @@
+that it's not always possible to port the quantum output of a system directly to hardware without
+first measuring it.
+
+In this tutorial, we will use PennyLane to do the following:
+
+1. Create an unknown target quantum process.
+2. Create initial states to feed into the target process.
+3. Measure the classical shadows of the target process.
+4. Create a model variational circuit to learn the quantum process.
+5. Train the variational circuit.
```
Since this is visible in the TOC, I recommend simplifying this into a sentence, rather than replicating it

DSGuala replied:
> replicating
Sorry, not sure what you mean by this?
I can rephrase this into one or two sentences though.

josh146 reviewed (commented):
Nice work @DSGuala --- I've left an initial round of comments. No need to tag me again for review, as @Qottmann and @ikurecic have this covered well :)

Qottmann commented:
There were merge conflicts with `master` which prevented building of the latest version. I made some local changes but that might mess with your poetry config. You can just reset hard to your last commit ([c77df90](https://github.com/PennyLaneAI/qml/pull/1156/commits/c77df90d07856fe5cba783a5b507e20a1ea71573)) and then fix poetry with a new lock from your environment.

Please make sure to review the last commit in the deployment (and make sure there is a deployment in the first place) when requesting re-review 👍

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,369 @@
+# be similar:
+#
+
+print("Untrained state for qubit 1:\n", model_circuit(initial_params, random_states[0])[1])
+
+print("Trained state for qubit 1:\n", model_circuit(params, random_states[0])[1])
+
+print("Target state for qubit 1:\n", pnp.mean(shadows[0].local_snapshots(), axis=0)[1])
```
TBH these dont look super convincing, on one hand there are readily visible differences (e.g. `0.12916712-0.25673602j` vs `0.00705-0.13455j` in the bottom left. And on the other hand, this is just comparing the local density matrices, which is just a small fraction of the total state. Generally, for comparing unitary matrices, I dont think comparing their numerical values is that insightfuly.

Instead, it would be good to show something that is more mathematically sound and physically motivated. The easiest thing to do would be taking the trace distance between the output unitaries of the model and target circuits. Since we are doing small sizes we can easily do that. There are other things that are more realistic to do in experiment, but I think this is easiest and most insightful

ikurecic replied:
Yeah, I agree... that would be better.
@DSGuala , either use a better way of comparison or kick out the whole section.

DSGuala replied:
I removed the section because the circuits that return classical shadows are not very compatible with `qml.matrix` and adding a matrix-able version of the circuit would make this demo even longer.

Alternatively, the density matrices now look like much closer and I could put them back:
```
Trained state for qubit 1:
 [[ 0.42897056+0.j         -0.0453594 +0.11168388j]
 [-0.0453594 -0.11168388j  0.57102944+0.j        ]]
Target state for qubit 1:
 [[ 0.42845+0.j      -0.0591 +0.09195j]
 [-0.0591 -0.09195j  0.57155+0.j     ]]
 ```

Qottmann replied:
Not optimal but fine for the sake of brevity for me 👍

Qottmann replied:
Note that you'd just need to redefine the target circuit without the classical shadows, as the model circuit can be done without re-defining anything. Would actually be nice to be able to do the same with the shadows circuit and just have qml.matrix ignore the return type since it isnt involved in the circuit matrix computation anyway

```
n = 4
params = np.ones(n)
random_state = np.random.rand(2**4)
random_state /= np.linalg.norm(random_state)

@qml.qnode(dev)
def model_circuit(params, random_state):
    # this is a parameterized quantum circuit with the same gate structure as the target Trotterized unitary
    qml.StatePrep(random_state, wires=range(n))
    for i in range(n):
        qml.RX(params[i], wires=i)

    return [qml.density_matrix(i) for i in range(n)]

qml.matrix(model_circuit, wire_order=range(n))(params, random_state)
```

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,369 @@
+#
+# .. note ::
+#
+#    We use few shadows and only one training state to keep the computational time low.
```
Do we expect good results with these restrictions?

DSGuala replied:
Not really. The cost should still reach a low point since it doesn't actually measure the overlap between the target and model circuits. But the Hilbert-Schmidt inner product of the target and model circuit matrices will not be optimal, nor will the trace distance between them.

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,369 @@
+
+######################################################################
+#
+# After training, the model outputs are similar to the target outputs estimated
```
That is quite surprising given these limitations no? Can this be?

DSGuala replied:
I think because we only trained on one input-output pair, it makes sense that we learned that input-output pair well. If we tried for any other pairs, it should look off. I can add this in for clarity.

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,369 @@
+#
+#     Sofiene Jerbi, Joe Gibbs, Manuel S. Rudolph, Matthias C. Caro, Patrick J. Coles, Hsin-Yuan Huang, Zoë Holmes
+#     "The power and limitations of learning quantum dynamics incoherently"
+#     `arXiv:2303.12834 <https://arxiv.org/abs/quant-ph/0512209>`__, 2005.
```
link points to https://arxiv.org/abs/quant-ph/0512209 lol 😅

DSGuala replied:
What! Fixed.

Qottmann reviewed (commented):
Great update @DSGuala 👍 

I think I understand the intuition behind that cost function now. Can you check if that makes sense? If yes, it would be good to include a small comment to convey that 👍 

From the provided data and numbers, I am still not sure a user can see whether the endeavour is successful or not. I think it would be good to plot the training curve and elaborate on the cost function minimum.

This brings me to my next point: how successful to we expect this whole thing to be in the first place? The paper is titled "The power **and limitations** of learning quantum dynamics incoherently", though here in this how-to it sounds a bit like this works with no problems and all is well - I really dont know the details, but is that the case? Or are there some limitations we should be aware of and convey?

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,369 @@
+#
+# .. note ::
+#
+#    On a personal computer, this method becomes slow (>1 second) around 10 qubits.
```
> Haar random vs non-Haar random
Since it's a toy problem and we are only using a few states, we could probably achieve the same results with non-Haar random states.
> *states* vs just using unitaries
That's an interesting suggestion, we could use `qml.QubitUnitary` with the Haar random unitaries instead of `qml.StatePrep` with the Haar random *states*

Qottmann replied:
I thought of something even easier, I vaguely recall seeing some simple product state construction for haar random inputs to these type of problems, but I might be mistaken, also cant find the reference now. So maybe just leave as is, sorry for the confusion ^^

DSGuala replied:
The paper itself says that stabilizer product states can be good enough to reproduce the results of Haar random states within some bound. But I think this is okay.

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,369 @@
+
+    for i in reversed(range(n_qubits - 1)):
+        qml.IsingZZ(params[n_qubits + i], wires=[i, i + 1])
+    return [qml.density_matrix(i) for i in range(n_qubits)]
```
Good point! I completely agree. The author's even allude to it.

> One could imagine here making use of Clifford [or Pauli]
shadows again to prepare states $$O^{(j)}(\theta)$$ on a quantum computer

But they also explain some disadvantages and then proceed to train classically.

Qottmann replied:
Yeah I think its fine like that, perhaps just a small comment because it might not be clear to every reader how one _measures_ the local density matrix (i.e. one would measure its paulis and reconstruct the denisty matrix from that, so tomography)

DSGuala replied:
Added a short sentence about this.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+# will approximate the time evolution of the corresponding transverse-field Ising system.
+#
+
+dev = qml.device("default.qubit")
+
+
+@qml.qnode(dev)
+def target_circuit(input_state):
+    # prepare training state
+    qml.StatePrep(input_state, wires=range(n_qubits))
+
+    # evolve according to desired hamiltonian
+    qml.TrotterProduct(hamiltonian, 2, 1, 1)
+    return qml.classical_shadow(wires=range(n_qubits))
+
+
+qml.draw_mpl(target_circuit)(random_states[0])
```
```suggestion
dev = qml.device("default.qubit")

@qml.qnode(dev)
def target_circuit(input_state):
    # prepare training state
    qml.StatePrep(input_state, wires=range(n_qubits))

    # evolve according to desired hamiltonian
    qml.TrotterProduct(hamiltonian, 2, 1, 1)
    return qml.classical_shadow(wires=range(n_qubits))

qml.draw_mpl(target_circuit)(random_states[0])
```
It looks odd to me if you start using two linebreaks here.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,369 @@
+# learns to produce the same output as the target circuit. We will then use the classical shadow
+# measurements to estimate the similarity between the ``model_circuit`` and the ``target_circuit``.
+#
+# As done in Jerbi et al. [#Jerbi]_, we create a ``model_circuit`` with the same gate structure as the target
```
Again, agreed here, @DSGuala .

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+# measurements to estimate the similarity between the ``model_circuit`` and the ``target_circuit``.
+#
+# As done in Jerbi et al. [#Jerbi]_, we create a ``model_circuit`` with the same gate structure as the target
+# structure. If the target quantum process were truly unknown, then we could choose a general
+# variational quantum circuit like in the :doc:`Variational classifier demo </demos/tutorial_variational_classifier>`.
```
```suggestion
# structure. If the target quantum process were truly unknown, then we would have beenable to choose a general
# :doc:`variational quantum circuit </demos/tutorial_variational_classifier>` instead.
```

DSGuala replied:
This suggestion might be a bit different from what I want to express.

The message is the following:
1. We know what the target circuit structure is, so we cheat a little and choose a variational circuit that looks like it.
2. If we *didn't* know what the target is, how would we choose a variational circuit?
3. We could go for a general one like in the demo.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+plt.show()
+
+######################################################################
+# 5. Training a model circuit using the classical shadows in a cost function
+# ----------------------------------------------------------------------------
```
```suggestion
# 5. Training a model circuit using classical shadows in a cost function
# ------------------------------------------------------------------------
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+# ----------------------------------------------------------------------------
+#
+# We now have to find the optimal parameters for ``model_circuit`` to mirror the ``target_circuit``.
+# We can estimate the similarity between the circuits according to the cost function provided in
+# appendix B of Jerbi et al. [#Jerbi]_.
```
```suggestion
# We can estimate the similarity between the circuits according to this cost function (see
# appendix B of [#Jerbi]_):
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+#
+# .. math:: C^l_N(\theta) = 1 - \frac{1}{nN}\sum^N_{j=1}\sum^n_{i=1}Tr[U|\psi^{(j)}\rangle\langle\psi^{(j)}|U^\dagger O^{(j)}_i(\theta)],
+#
+# Where :math:`n` is the number of qubits, :math:`N` is the number of initial states, :math:`\psi^{(j)}`
```
```suggestion
# where :math:`n` is the number of qubits, :math:`N` is the number of initial states, :math:`\psi^{(j)}`
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+# are random states, :math:`U` is our target unitary operation, and :math:`O_i` is the local density
+# matrix for qubit :math:`i` after applying the ``model_circuit``.
+#
+# We can calculate this cost for our system by using the shadow measurements made before to estimate
```
```suggestion
# We can calculate this cost for our system by using the `shadow measurements <#time-evolution-and-classical-shadow-measurements>`_ to estimate
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+#
+# We can calculate this cost for our system by using the shadow measurements made before to estimate
+# the expectation value of :math:`O_i`. This gives an estimate of the overlap between the state of
+# the target circuit and the model circuit:
```
```suggestion
# the target circuit and the model circuit.
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,369 @@
+# .. math:: C^l_N(\theta) = 1 - \frac{1}{nN}\sum^N_{j=1}\sum^n_{i=1}Tr[U|\psi^{(j)}\rangle\langle\psi^{(j)}|U^\dagger O^{(j)}_i(\theta)],
+#
+# Where :math:`n` is the number of qubits, :math:`N` is the number of initial states, :math:`\psi^{(j)}`
+# are random states, :math:`U` is our target unitary operation, and :math:`O_i` is the local density
```
I think this commentary is fine to be left out. This is a How-to PennyLane Demo, showing an implementation via PennyLane, and not an elaboration of the paper. If anyone is confused about anything, they'll go to the paper and/or authors.

DSGuala replied:
>Roughly, this cost function measures the fidelity between the model circuit and the target circuit by proxy of the single-qubit reduced states $\rho_{i}^{(j)} =: O_{i}^{(j)}(\theta)$ of the target over a variety of input-output pairs."

Added a short explanation for this using that sentence. Let me know if it's too much.

Qottmann replied:
looks good to me 👍

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+# Using the learning dynamics incoherently dataset
+# ----------------------------------------------------------------
+#
+# In Jerbi et al. [#Jerbi]_, the authors perform the procedure described above on a larger, 16-qubit transverse field Ising
```
```suggestion
# In Jerbi et al. [#Jerbi]_, the authors perform this procedure to learn dynamics incoherently on a larger, 16-qubit transverse-field Ising
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+# ----------------------------------------------------------------
+#
+# In Jerbi et al. [#Jerbi]_, the authors perform the procedure described above on a larger, 16-qubit transverse field Ising
+# Hamiltonian and uses classical shadow samples from quantum hardware to estimate the cost function.
```
```suggestion
# Hamiltonian, and use classical shadow samples from quantum hardware to estimate the cost function.
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+#
+# In Jerbi et al. [#Jerbi]_, the authors perform the procedure described above on a larger, 16-qubit transverse field Ising
+# Hamiltonian and uses classical shadow samples from quantum hardware to estimate the cost function.
+# This data is available in PennyLane through the :mod:`qml.data` module. More information about
+# the dataset itself is available on the
+# `Learning Dynamics Incoherently dataset page <https://pennylane.ai/datasets/other/learning-dynamics-incoherently>`_.
```
```suggestion
# These data are available in `PennyLane Datasets <https://pennylane.ai/datasets/>`__ through the :mod:`qml.data` module, as the 
# `Learning Dynamics Incoherently dataset <https://pennylane.ai/datasets/other/learning-dynamics-incoherently>`_.
```
Let's keep it concise.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+# the dataset itself is available on the
+# `Learning Dynamics Incoherently dataset page <https://pennylane.ai/datasets/other/learning-dynamics-incoherently>`_.
+#
+# We first load the dataset using :func:`~pennylane.data.load`:
```
```suggestion
# To use this dataset, we first load it using :func:`~pennylane.data.load`:
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+
+######################################################################
+#
+# The unknown target hamiltonian, Haar-random intial states, and resulting classical shadow
```
```suggestion
# The unknown target Hamiltonian, Haar-random intial states, and resulting classical shadow
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,372 @@
+######################################################################
+#
+# We only need to create the model circuit, cost function, and train.
+# For these we use the same model circuit as above, updated to reflect the increased number
```
```suggestion
# For these we use the same model circuit as `above <#creating-a-model-circuit-that-will-learn-the-target-process>`_, updated to reflect the increased number
```

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,293 @@
+state is to the target process output, we use the classical shadows to estimate the overlap of the
+states. [double check wording here, we aren't exactly estimating the overlap]
+
+This is different to learning the quantum process *coherently* because it does not require the model
+circuit to be connected to the target quantum process. That is, the model circuit does not receive
+quantum information from the target process directly. Instead, we train the model circuit using
+classical information obtained from the classical shadow measurements. One reason this is useful is
+that it's not always possible to port the quantum output of a system directly to hardware without
+first measuring it.
```
perhaps this https://arxiv.org/abs/2112.00778 , there is probably a newer one but I quite liked this one

DSGuala replied:
Added :+1:

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,346 @@
+r"""How to learn quantum dynamics incoherently
```
Yup, 100% agree, let's change the title. Not sure what the best version is yet though. But I would like to convey as much as possible from these four points:

1. Reproducing an unknown quantum process
2. This is done by optimizing a variational circuit
3. We use classical shadows
4. We don't need the output of the unknown process to go directly into the variational circuit

Maybe:
- Too long: "How to learn a quantum process from classical shadow measurements using a variational circuit"
- A bit awkward: "How to variationally learn a quantum process using classical shadows"
- Missing the variational part: "How to reproduce a quantum process using classical shadows"

Qottmann replied:
I think actually this one is the most apt

> "How to variationally learn a quantum process using classical shadows"

josh146 replied:
The one advantage to having the title closely match the paper, however, is that researchers searching for the paper will also find our demo/how to.

This ends up being a big reason our content is discovered, so I recommend preserving the article title if possible

ikurecic replied:
```suggestion
r"""Learning quantum dynamics incoherently: Variational learning using classical shadows
```
How about this?

josh146 replied:
That is great!

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,375 @@
+#    We will be performing *local* measurements to keep the computational complexity lower and
+#    because classical shadows are well-suited to estimating local observables [#Jerbi]_.
+#    For this reason, the following circuit returns local density matrices for each qubit. In
+#    hardware, the density matrix is obtained via Pauli measurements and state tomography.
```
```suggestion
#    hardware, the density matrix is obtained via state tomography using Pauli measurements of classical shadows.
```

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,385 @@
+# ----------------------------------------------
+#
+# For our unknown quantum process, we will use a well-known quantum process,
+# the :doc:`time evolution of a Hamiltonian </demos/tutorial_qaoa_intro/#circuits-and-hamiltonians>`:
```
> tutorial_qaoa_intro/#circuits-and-hamiltonians
@ikurecic This doesn't seem to work :thinking:
The only other example I've found is in the [lie simulation demos](https://pennylane.ai/qml/demos/tutorial_liesim/) but even there it doesn't seem to work as expected.
Okay to go back to linking wikipedia?

ikurecic replied:
Interesting, that seems like a bug — thanks, @DSGuala , I'll report it. 
In the meantime, please link to the entire Demo.

DSGuala replied:
Updated

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.metadata.json:
```diff
@@ -0,0 +1,46 @@
+        }
+    ],
+    "dateOfPublication": "2024-07-09T00:00:00+00:00",
+    "dateOfLastModification": "2024-07-09T00:00:00+00:00",
```
This was updated to 07-29

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.metadata.json:
```diff
@@ -0,0 +1,46 @@
+    "tags": [],
+    "previewImages": [
+        {
+            "type": "thumbnail",
+            "uri": "/_static/demonstration_assets/regular_demo_thumbnails/thumbnail_learning_dynamics_incoherently.png"
+        },
+        {
+            "type": "large_thumbnail",
+            "uri": "/_static/large_demo_thumbnails/thumbnail_large_learning_dynamics_incoherently.png"
+        }
```
Hey @ikurecic , not sure what the issue is here?

ikurecic replied:
Some incoming changes that I can handle in the last step, as described on Slack. You can leave this to me, but the 'tk' is here as a reminder. Thanks!

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+# 
+# More sppecifically, we will approximate :math:`U(H, t)` via
+# `Trotterization <https://en.wikipedia.org/wiki/Hamiltonian_simulation#Product_formulas>`_.
+# We first create the Hamiltonian and Trotterize later with :class:`~pennylane.TrotterProduct`.
```
```suggestion
# We first create the Hamiltonian and trotterize later with :class:`~pennylane.TrotterProduct`.
```

not sure

DSGuala replied:
Gave this a Google search :+1:  seems to be "Trotterize" because Trotter is a last name.

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+pnp.random.seed(7)
+
+# generate random weights
+alphas = pnp.random.normal(0, 0.5, size=n_qubits)
```
do you actually need to use pnp for coefficients you wont be differentiating? I _think_ you can just use regular numpy here, should be faster as well

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+    # prepare training state
+    qml.StatePrep(input_state, wires=range(n_qubits))
+
+    # evolve according to desired Hamiltonian
+    qml.TrotterProduct(hamiltonian, 2, 1, 1)
```
```suggestion
    # evolve the Hamiltonian for t=2 in n=1 steps with the order 1 formula
    qml.TrotterProduct(hamiltonian, t=2, n=1, order=1)
```

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+# As done in [#Jerbi]_, we create a model circuit with the same gate structure as the target
+# circuit. If the target quantum process were truly unknown, then we would choose a general
+# variational quantum circuit like in the
+# :doc:`Variational classifier demo </demos/tutorial_variational_classifier>`.
```
link doesnt work in the deployment, which is odd 🤔 cant see why

DSGuala replied:
Turns out this is because the other demos aren't built in the PR preview :+1:

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+#
+# We can calculate this cost for our system by using the
+# `shadow measurements <#time-evolution-and-classical-shadow-measurements>`_ to estimate
+# the expectation value of :math:`O_i`. Roughly, this cost function measures the fidelity between
+# the model circuit and the target circuit, by proxy of the single-qubit reduced states
+# :math:`\rho_{i}^{(j)}` of the model over a variety of input-output pairs.
```
👍

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+
+
+# Find the ideal parameters from the original Trotterized Hamiltonian
+ideal_parameters = [
```
cool!

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+#
+# .. math:: Tr[(\rho_{i}^{(j)})^2] < 1.
+#
+# The ideal cost :math:`C^l_N(\theta)` is therefore greater than 0.
```
nice explainer!

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+)
+model_matrix = qml.matrix(model_circuit, wire_order=range(n_qubits))(params, random_states[0])
+
+trace_distance = qml.math.trace_distance(target_matrix, model_matrix)
```
Thats cool!

Two things: The random input states are a means to make the incoherent learning work with classical shadows. But now we want to compare just the unitaries of the circuit themselves. In order to not have to re-define the circuits without the state prep, you can just input the computational 0 basis state (`np.eye(2**n)[0]`), which in the matrix calculation should decompose to the identity.

Second, because the number is relatively large still, this doesnt tell us too much at this point. An even better way to showcase this is by showing how the trace distance changed over the course of the optimizaiton.

At the risk of being really annoying I can provide you with the necessary code, feel free to include it or discard at your discretion, I dont want to overburden you too much.

(see the next few code commit suggestions, marked as "optional code suggestion for trace distance x/N")

Qottmann replied:
You can compare it to training cost and accuracy in Machin Learning: the cost (here also cost) is a proxy to optimize accuracy (here fidelity), which is harder to directly target

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+    @ qml.StatePrep(random_states[0], wires=range(n_qubits)),
+    wire_order=range(n_qubits),
+)
+model_matrix = qml.matrix(model_circuit, wire_order=range(n_qubits))(params, random_states[0])
```
```suggestion
model_matrices = [qml.matrix(model_circuit, wire_order=range(n_qubits))(pp, random_states[0]) for pp in paramss]
```

optional code suggestion for trace distance 1/3

Qottmann replied:
note that the code suggestion also includes the `np.eye(2**n)` suggestion which I think should definitely be part of the demo

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+steps = 50
+costs = [None] * (steps + 1)
+costs[0] = cost(initial_params)
+
+for i in range(steps):
+    params, costs[i + 1] = optimizer.step_and_cost(cost, params)
```
```suggestion

paramss = [] # keep track of parameters during optimization
for i in range(steps):
    params, costs[i + 1] = optimizer.step_and_cost(cost, params)
    paramss.append(params)
```

optional code suggestion for trace distance 2/3

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+)
+model_matrix = qml.matrix(model_circuit, wire_order=range(n_qubits))(params, random_states[0])
+
+trace_distance = qml.math.trace_distance(target_matrix, model_matrix)
```
```suggestion
trace_distances = [qml.math.trace_distance(target_matrix, model_matrix) for model_matrix in model_matrices]

plt.plot(trace_distances)
```

optional code suggestion for trace distance 3/3

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+
+######################################################################
+#
+# For this number of qubits it becomes prohibitive to obtain the matrix and calculate the trace
+# distance. For a sanity check, we can instead take a look at the density matrices
```
```suggestion
# As a quick check, we can take a look at the density matrices
```

For 16 qubits it shouldnt be prohibitive to compute the trace distance, but its okay not to do it here 👍

DSGuala replied:
Thanks. I agree and made the change.

To make sure we're on the same page, is it just that "prohibitive" is too strongly worded? 

What I'm trying to say is that creating the 2^16 by 2^16 matrix takes up a lot of memory. We can do some clever manipulation to still calculate the trace distance. 

But we will have to put in some more effort or more memory. Otherwise:
![image](https://github.com/user-attachments/assets/453aa729-c8c8-48e0-9e6f-06bddb5f25d8)

Qottmann replied:
yep was just about prohibitive being too strong :)

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+######################################################################
+#
+# After training, the model outputs are closer to the target outputs, but not quite the same.
+# This can be improved by using more training states and classical shadow measurements.
```
```suggestion
# This can be improved by using more training states and classical shadow measurements.
# We see the limitations of incoherent learning as even for an easy unitary circuit of just a single Trotter step we need a good amount of resources to faithfully reproduce the underlying quantum process.
```

Doesnt have to be this sentence exactly but I think it would be important to stress the limitations again, which are quite evident from the so-so performance

Qottmann reviewed (approved):
Great job @DSGuala 

Happy to approve the demo from the technical side now :)

I left a bunch of nitpicky comments and suggestions, some of which are optional and some of which I consider would improve the demo. Feel free to include and discard those marked as such at your discretion :)

Qottmann commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+
+target_matrix = qml.matrix(
+    qml.TrotterProduct(hamiltonian, 2, 1, 1)
+    @ qml.StatePrep(random_states[0], wires=range(n_qubits)),
```
```suggestion
    @ qml.StatePrep(np.eye(2**n), wires=range(n_qubits)),
```

DSGuala replied:
I simply removed the state prep part since it doesn't change this matrix.

Qottmann replied:
yes in the matrix one that is expected, I guess in the model circuit you cant just remove it so I left it so make the analogy clear. I also missed the `[0]` in the suggestion here 😬

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+# ----------------------------------------------
+#
+# For our unknown quantum process, we will use the
+# `time evolution of a Hamiltonian <https://pennylane.ai/qml/demos/tutorial_qaoa_intro/#circuits-and-hamiltonians>`_:
```
We don't really have a better one that I've seen.
The other option is to link to wikipedia

Qottmann replied:
then perhaps just remove the link 👍

ikurecic replied:
Let's leave that decision to the Product Team in the final check. (CC @justinpickering , please take a look at this comment once this Demo is ready for the final check.)

justinpickering replied:
Since we don't have any better options, we can remove the link.
```suggestion
# time evolution of a Hamiltonian:
```

ikurecic replied:
Thank you, Justin. @dsguala , it would be good if you could remove it.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.metadata.json:
```diff
@@ -0,0 +1,65 @@
+{
+    "title": "Learning Dynamics Incoherently",
```
```suggestion
    "title": "Learning dynamics incoherently",
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.metadata.json:
```diff
@@ -0,0 +1,65 @@
+            "id": "diego_guala"
+        }
+    ],
+    "dateOfPublication": "2024-07-29T00:00:00+00:00",
+    "dateOfLastModification": "2024-07-29T00:00:00+00:00",
```
TK

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.metadata.json:
```diff
@@ -0,0 +1,65 @@
+    "previewImages": [
+        {
+            "type": "thumbnail",
+            "uri": "/_static/demonstration_assets/regular_demo_thumbnails/thumbnail_learning_dynamics_incoherently.png"
```
@DSGuala , do you mind if I commit to your branch to fix this path? There was an update earlier this week.

DSGuala replied:
Yup, no problem!

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.metadata.json:
```diff
@@ -0,0 +1,65 @@
+        },
+        {
+            "type": "large_thumbnail",
+            "uri": "/_static/large_demo_thumbnails/thumbnail_large_learning_dynamics_incoherently.png"
```
TK, same as with the other thumbnail.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.metadata.json:
```diff
@@ -0,0 +1,65 @@
+            "uri": "/_static/large_demo_thumbnails/thumbnail_large_learning_dynamics_incoherently.png"
+        }
+    ],
+    "seoDescription": "Learning Dynamics Incoherently",
```
Could you please add a sentence here to 'pitch' and describe the Demo instead of just the title?

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+r"""Learning quantum dynamics incoherently: variational learning using classical shadows
+====================================================================================
```
```suggestion
r"""Learning quantum dynamics incoherently: Variational learning using classical shadows
========================================================================================
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+in the
+`Learning Dynamics Incoherently PennyLane Dataset <https://pennylane.ai/datasets/other/learning-dynamics-incoherently>`__. 
+
+In simple terms, learning dynamics incoherently consists of two steps. First, we measure the output
+of the unknown process for many different inputs. Then, we adjust a variational quantum circuit
+until it produces the same input-output combinations as the unknown process. For step 1, we measure
+:doc:`classical shadows <tutorial_classical_shadows>` of the target process output.
+For step 2, we simulate the model circuit to get its final state. To know how similar the simulated
+state is to the target process output, we estimate the overlap of the states using the classical
+shadows from step 1.
```
I'm confused reading this paragraph for the first time. Is the "first" actually "step 1" or are there two steps in the "first?" Maybe a slight rephrasing would be helpful.

DSGuala replied:
Rephrased :)

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+For step 2, we simulate the model circuit to get its final state. To know how similar the simulated
+state is to the target process output, we estimate the overlap of the states using the classical
+shadows from step 1.
+
```
This is a pretty long introductor session. Maybe you'd want to put the OpenGraph image somewhere between two paragraphs to break it up a bit?

DSGuala replied:
The section is a bit shorter now, do we still want the OpenGraph image?

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.metadata.json:
```diff
@@ -0,0 +1,65 @@
+    "dateOfPublication": "2024-07-29T00:00:00+00:00",
+    "dateOfLastModification": "2024-07-29T00:00:00+00:00",
+    "categories": [
+        "Quantum Machine Learning"
```
```suggestion
        "Quantum Machine Learning",
        "How-to"
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,421 @@
+first measuring it. However, depending on the quantum process, an exponential number of classical
+shadow measurements may be required [#Jerbi]_.
+
+In this tutorial, we will use PennyLane to create an unknown target quantum process and the initial
+states to go into it, simulate shallow classical shadow measurements of the target process, and
+create and train a model variational circuit that approximates the target process.
+Finally, we will use the
+`Learning Dynamics Incoherently PennyLane Dataset <https://pennylane.ai/datasets/other/learning-dynamics-incoherently>`__
+to replicate part of the original investigation.
```
I partially agree with Korbinian here. Removing it does not lose any information, but maybe it would be better to remove it and then swap the second nad third paragraphs (after my comment above is addressed). That way there would be a more compelling pitch than is currently at the end of the third paragraph.

We might need one final tweak after that's done, to make the opening interesting.

DSGuala replied:
Paragraph removed and other two swapped with some small tweaks. Let me know your thoughts.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+# behavior for states that are very different from our training set.
+#
+# For quantum systems, this means we want to sample
+# :doc:`Haar random states <tutorial_haar_measure>` as done below.
```
```suggestion
# :doc:`Haar-random states <tutorial_haar_measure>`, as done below.
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+#
+# Since ``target_circuit`` returns :func:`~pennylane.classical_shadow`, running the circuit with a
+# ``shot`` value gives the desired number of classical shadow measurements.
+# We use this to create a set of shadows for each initial state:
```
```suggestion
# We use this to create a set of shadows for each initial state.
```
For consistency.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+# As done in [#Jerbi]_, we create a model circuit with the same gate structure as the target
+# circuit. If the target quantum process were truly unknown, then we would choose a general
+# variational quantum circuit like in the
+# :doc:`variational classifiers demo </demos/tutorial_variational_classifier>`.
```
```suggestion
# :doc:`Variational classifier demo </demos/tutorial_variational_classifier>`.
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+@qml.qnode(dev)
+def model_circuit(params, random_state):
+    qml.StatePrep(random_state, wires=range(n_qubits))
+    # Parameterized quantum circuit with the same gate structure as the target
```
```suggestion
    # parameterized quantum circuit with the same gate structure as the target
```
For consistency.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+plt.show()
+
+######################################################################
+# 5. Training a model circuit using classical shadows in a cost function
+# ------------------------------------------------------------------------
```
```suggestion
# 5. Training using classical shadows in a cost function
# ------------------------------------------------------
```
This is a two-line heading otherwise.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+def cost(params):
+    cost = 0.0
+    for idx, random_state in enumerate(random_states):
+        # Obtain the density matrices for each qubit
```
```suggestion
        # obtain the density matrices for each qubit
```
For consistency.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+    for idx, random_state in enumerate(random_states):
+        # Obtain the density matrices for each qubit
+        observable_mats = model_circuit(params, random_state)
+        # Convert to a PauliSentence
```
```suggestion
        # convert to a PauliSentence
```
For consistency.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+            qml.pauli_decompose(observable_mat, wire_order=[qubit])
+            for qubit, observable_mat in enumerate(observable_mats)
+        ]
+        # Estimate the overlap for each qubit
```
```suggestion
        # estimate the overlap for each qubit
```
For consistency. (Note that you can either change these or capitalize other comments in previous codeblocks.)

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+#
+
+
+# Find the ideal parameters from the original Trotterized Hamiltonian
```
```suggestion
# find the ideal parameters from the original Trotterized Hamiltonian
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+plt.ylabel("Cost")
+plt.xlabel("Training iterations")
+plt.legend()
+plt.show()
```
It might be nice to label it as "Ideal parameters" instead of "Ideal Parameters."

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+#
+# The ideal cost :math:`C^l_N(\theta)` is therefore greater than 0.
+#
+# We can also look at the :func:`trace distance <pennylane.math.trace_distance>` between the unitary
```
```suggestion
# We can also look at the :func:`trace_distance <pennylane.math.trace_distance>` between the unitary
```
Like this?

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+
+
+######################################################################
+# Using the learning dynamics incoherently dataset
+# ----------------------------------------------------------------
```
```suggestion
# Using the Learning Dynamics Incoherently dataset
# ------------------------------------------------
```
For consisntency

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+# In Jerbi et al. [#Jerbi]_, the authors perform this procedure to learn dynamics incoherently on a
+# larger, 16-qubit transverse-field Ising
+# Hamiltonian, and use classical shadow samples from quantum hardware to estimate the cost function.
+# This data is available in the
```
```suggestion
# These data are available in the
```
Alternatively "This dataset is …"

DSGuala replied:
Used
> The corresponding [Learning Dynamics Incoherently PennyLane Dataset](https://pennylane.ai/datasets/other/learning-dynamics-incoherently) can be downloaded via the `qml.data` module.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+# We only need to create the model circuit, cost function, and train.
+# For these we use the same model circuit as
+# `above <#creating-a-model-circuit-that-will-learn-the-target-process>`_, updated to reflect
+# the increased number of qubits:
```
```suggestion
# the increased number of qubits.
```

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,437 @@
+# About the author
+# ------------------
+#
+# .. include:: ../_static/authors/diego_guala.txt
```
```suggestion
```
Superfluous.

ikurecic reviewed (commented):
Thanks, @DSGuala . Here are some suggestions.

justinpickering commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,429 @@
+#
+# We can then minimize the cost to train the model to output the same states as the target circuit.
+# For this, we can use the cost function from
+# `before <#training-a-model-circuit-using-classical-shadows-in-a-cost-function>`_, 
```
The `before` link was broken, I believe this should correctly link to this section?
```suggestion
# `before <#training-using-classical-shadows-in-a-cost-function>`_, 
```

justinpickering reviewed (approved):
Great work @DSGuala! Just a couple small comments/suggestions below.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,429 @@
+in the
+`Learning Dynamics Incoherently PennyLane Dataset <https://pennylane.ai/datasets/other/learning-dynamics-incoherently>`__. 
+
+This approach differs from learning the quantum process *coherently* [#Huang]_ because it does not
```
This reference is new. I'm not sure I understand why it's here. Could you please add more context or remove it?

Qottmann replied:
I requested this, I dont think it is widely known what is meant with "coherently" and "incoherently" learning from a quantum experiment, this reference is here to showcase what the "coherent" case looks like (and the text specifies that this is cases where the model and system to study can share quantum info, whereas in the incoherent case studied here we need to use measurements as a proxy to connect model and target system

DSGuala replied:
I tried a few different ways to make it clear that the reference is proposing a coherent learning procedure, but ultimately felt like the current version works best :+1:

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,429 @@
+of the unknown process for many different inputs. For example, we will measure
+:doc:`classical shadows <tutorial_classical_shadows>` of the target process output.
+Then, we adjust a variational quantum circuit
+until it produces the same input-output combinations as the unknown process. In this tutorial, we
```
People use this term in a conversational sense, as opposed to a definition when it is used within the text body non-capitalized. I believe that this is fine. However, I would suggest a new paragraph here, if you could do that, @DSGuala  .

I also think the OpenGraph image is missing from the opening. Didn't we say this would be added? I have been on holiday for a week, so I might be misremembering, but this is a lot of text for an opening to not be broken up by a visual. I can add it here if you tell me if you have any preferences.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,429 @@
+
+
+######################################################################
+# Using the Learning Dynamics Incoherently dataset
+# ------------------------------------------------
```
Could you title it _Using the Learning Dynamics Incoherently PennyLane Dataset_? We should promote this accurately.

ikurecic commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,429 @@
+# short-time evolution of a first order single Trotter step, it requires a large number of
+# shadow measurements and training states to faithfully reproduce the underlying quantum process.
+# The results can be improved by increasing the number training states and
+# classical shadow measurements.
```
This is probably a good place to link the Demo again.

ikurecic reviewed (commented):
@DSGuala , there are some very minor comments left by Justin from yesterday and from me today (typos and link/reference suggestions) after the recent changes you've made.
If you could address those, I think we can merge and deploy, if everything runs as expected at that point.

DSGuala commented on demonstrations/tutorial_learning_dynamics_incoherently.py:
```diff
@@ -0,0 +1,429 @@
+#
+# where :math:`n` is the number of qubits and :math:`\alpha` are randomly generated weights.
+# 
+# More sppecifically, we will approximate :math:`U(H, t)` via
```
Nice catch!

justinpickering commented on demonstrations/tutorial_learning_dynamics_incoherently.metadata.json:
```diff
@@ -0,0 +1,66 @@
+    "title": "Learning dynamics incoherently",
+    "authors": [
+        {
+            "id": "diego_guala"
```
Note, will need to update this in metadata as the profile linking PRs have been merged https://github.com/XanaduAI/pennylane.ai-react/pull/1618 and https://github.com/PennyLaneAI/qml/pull/1181
```suggestion
            "username": "Diego"
```
